core/templates/dev/head/domain/collection/GuestCollectionProgressService.js

```diff
@@ -0,0 +1,146 @@
+      }
+      return store[collectionId];
+    };
+
+    var setCompletedExplorationIds = function(collectionId, explorationIds) {
+      var store = getCollectionProgressStore();
+      store[collectionId] = explorationIds;
+      setCollectionProgressStore(store);
+    };
+
+    var recordCompletedExploration = function(collectionId, explorationId) {
+      var completedIds = getCompletedExplorationIds(collectionId);
+      // Only register the exploration ID if it's not already in the list of
+      // completed IDs (to avoid duplication).
+      if (completedIds.indexOf(explorationId) === -1) {
+        completedIds.push(explorationId);
+        setCompletedExplorationIds(collectionId, completedIds);
+      }
+    };
+
+    var getValidCompletedExplorationIds = function(collection) {
+      var collectionId = collection.getId();
+      // Filter the exploration IDs by whether they are contained within the
+      // specified collection structure.
+      return getCompletedExplorationIds(collectionId).filter(function(expId) {
+        return collection.containsCollectionNode(expId);
+      });
+    };
+
+    // This method corresponds to collection_domain.get_next_exploration_ids.
+    var _getNextExplorationIds = function(collection, completedIds) {
+      // Given the completed exploration IDs, compile a list of acquired skills.
+      var acquiredSkillIds = completedIds.map(function(expId) {
+        var collectionNode = collection.getCollectionNodeByExplorationId(expId);
+        return collectionNode.getAcquiredSkillIds();
+      }).reduce(function(accumulator, value) {
+        return accumulator.concat(value);
+      }, []);
+
+      // Find remaining collection nodes which have yet to be completed.
+      var collectionNodes = collection.getCollectionNodes();
+      var incompleteNodes = collectionNodes.filter(function(node) {
+        return completedIds.indexOf(node.getExplorationId()) === -1;
+      });
+
+      // Filter out nodes for which the guest does not have the prerequisite
+      // skills for, then map the leftover nodes to their corresponding
+      // exploration IDs.
+      // https://stackoverflow.com/a/15514975
+      return incompleteNodes.filter(function(node) {
+        return node.getPrerequisiteSkillIds().every(function(elem) {
+          return acquiredSkillIds.indexOf(elem) > -1;
+        });
+      }).map(function(node) {
+        return node.getExplorationId();
+      });
+    };
+
+    return {
+      /**
+       * Records that the specified exploration was completed in the context of
+       * the specified collection, as a guest.
+       */
+      recordExplorationCompletedInCollection: function(
+          collectionId, explorationId) {
+        recordCompletedExploration(collectionId, explorationId);
+      },
+
+      /**
+       * Returns whether the guest user has made any progress toward completing
+       * the specified collection.
+       */
+      hasMadeProgress: function(collectionId) {
```

Comment:
Perhaps hasCompletedSomeExploration is better

Reply:
Sure; done.